### PR TITLE
Support restore point based snapshot isolation for hot standby

### PR DESCRIPTION
**Pull request by huansong**
Currently, the transaction isolation supported on hot standby has a number of key limitations:

1. It is possible that different standby segments have replayed to different WAL records at a given time (for example, seg0 replays an INSERT, but seg1 hasn't). So the query on standby could easily break ACID.
2. As one of the query conflict types for hot standby, the snapshot conflict in GPDB is not taking distributed snapshot into account, causing potential correctness issues. See isolation2/hot_standby/queryconflict test for example.
3. The hot standby is not able to view the last one-phase commit (1PC) visible, even if the standby segments has replayed them. This is because since 1PC does not have any WAL on QD, the standby QD won't advance its latestCompletedGxid, so its distributed snapshot horizon will not include the last 1PC - it would view them not-yet-started or at best, still-in-progress.
4. The primary segments truncate the distributed log that is no longer used by any distributed snapshots *on the primary nodes*. However, it does not account for distributed snapshot on the standby nodes. As a result, the standby which has truncated distributed log that it still needs might fail to do dtx visibility checks, and give wrong results.

We now use a restore point based approach to address all the above issues. The basic idea behind is that, we no longer allow the standby to create snapshots at random state. Instead, every standby transaction will be using snapshots that are created at certain restore point, which by design guaranttees that every segment should have consistent state of transactions.

The basic mechanism is:
* The primary creates restore points (RP) using gp_create_restore_point(), preferably in some periodic fashion (e.g. every 5 minutes).
* The standby coordinator/segment replays the RP WALs. Upon replaying each RP WAL, every standby coordinator/segment exports a snapshot, and remembers the RP name.
* When a query comes, the standby coordinator chooses an RP to use (either specified by the user, or just using the latest RP). Then, it imports a snapshot corresponding to that RP which will be used for the entirety of the transaction.
* When the standby coordinator dispatches the query, it includes the RP name in the dispatch message. Similar to what happened on the coordinator, the standby segmetns will import a snapshot corresponding to that RP. With this design, we no longer support read-committed isolation. It will always be repeatable-read, since the snapshot doesn't change in a transaction.

A simple work flow can be shown as:
figure

Here is the basic usage on standby:
```sql
-- Set the "snapshot mode" to restore-point based. In fact, this is the default as we would like to encourage the proper usage.
-- Other choise is 'unsync' which would be running w/o this approach (and give possibly inconsistent result).
set gp_hot_standby_snapshot_mode = restorepoint;

-- Set the restore point name we want to use. Or, leave it blank so we will be using the latest RP.
-- It will be recommended in our doc to also set synchronous_commit=remote_apply when using the latest RP.
set gp_hot_standby_snapshot_restore_point_name = 'rp1';

-- Run queries
select * from t;
```

With this approach, all of the limitations mentioned before will be addressed:
1. By creating snapshot and remembering RP at the WAL replay time, we make sure that all standby nodes have replayed sufficient WALs to proceed.
2. Because the fact that an RP is created while holding TwophaseCommitLock, it can be understood that a dtx must be either all committed locally, or all in-progress locally on all segments. That means, we no longer need to account for distributed snapshots for snapshot conflict - local snapshot conflict would provide the exact same result.
3. We now log the latestCompletedGxid with every RP on the primary (XLOG_LATESTCOMPLETED_GXID). Therefore, the RP-based snapshot on standby can correctly account for the 1PC that completed before the RP.
4. If distributed log is truncated while the standby segment still can see the corresponding tuple, what would happen is that the standby segment would resort to local visibility check because it cannot find the gxid for the tuple. Then, similar to the point 2 above, local visibility check will be enough to give the correct result.

P.S. an interesting insight from Ashwin is that, with this approach, we might no longer even need all the dtx snapshot checks in the MVCC, which might greatly help performance. We'll explore that possibility at a later point.

Dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/hsd-rp

plan for hot standby support in GPDB:
1. Enable hot standby dispatch on a mirrored cluster  [[done](https://github.com/greenplum-db/gpdb/pull/16817)]
2. Support dtx snapshot isolation for hot standby [[done](https://github.com/greenplum-db/gpdb/pull/17004)]
3. Handle hot standby query conflict [[done](https://github.com/greenplum-db/gpdb/pull/17004)]
4. Enable restore point based snapshot isolation [in-progress]
6. Support snapshot conflict for AO/CO table and index on hot standby [not started]

### CODE REVIEW

**Review by soumyadeep2007 — COMMENTED** [review]
First pass.

I think it may be possible to eliminate some of the deviations in the upstream snapmgr.c.

Here is what I'm thinking:

* We would like to continue to use the same file format and directory location (no prefix, string manipulation etc) for storing snapshots. Using a separate dir is appealing, but that might be cumbersome given all of our utilities would have to know of it.
* **Have a notion of a `persistent` exported snapshot, that outlives the transaction context that exports it.**
    * One way to achieve that is to have a `persistent_mainfest` flat file inside `pg_snapshots`, that lists all such snapshots.
    * We can modify `ExportSnapshot` to take a flag `persistent`, which if `true` would add the snapshot to the manifest. We can also add a `persistent` flag to the `ExportedSnapshot` struct, which if `true`, would cause `AtEOXact_Snapshot` to ignore the snapshot while cleaning up at end-of-xact.
    * `DeleteAllExportedSnapshotFiles()` can look at the persistent manifest and ignore those snapshots that are listed in the persistent manifest. If there is no manifest, it can go and nuke everything.
    * We would now have to remember the name of the snapshot in the hash table from now on, which is fine.
* We would need to have another manifest file to store the restore point -> snapshot mapping, so that we can rebuild the restore point hash table during startup. We could in theory make this info a part of the persistent_manifest, but that would make the concept of persistent snapshots coupled to restore points. (I'm not super sold on requiring this level of decoupling, but it might be nice..have to think more)

**Review by huansong — COMMENTED** [review]
> We would like to continue to use the same file format and directory location (no prefix, string manipulation etc)

We cannot use the current format, as (1) there's no valid xid for the startup process who creates the snapshot; (2) the name won't be unique across server restart (as backend ID would repeat and the snapshot counter starts from 1 again). So some kind of special format is needed, but we can discuss more how it should look like.

> Have a notion of a persistent exported snapshot, that outlives the transaction context that exports it.

I like that notion. Though I think that won't help us eliminate the deviation from upstream code. We would basically use a `persistent` flag instead of being given a `snapshot_name` - the footprints would be similar. And since we still need a special file name format as mentioned above, we would basically need to process two new parameters which feel unnecessary. 

Currently, more deviations are actually about the startup process not having transaction environment - that doesn't change with the notion.

> We can modify ExportSnapshot to take a flag persistent, which if true would add the snapshot to the manifest.

We could do that. Right now we actually can do very similar thing since we've added a `isRestorePointBased` flag to `SnapshotData`. However, that feels a bit over-engineering to me - the startup process would have to remember those snapshots in `ExportedSnapshot`, but it never reaches `AtEOXact_Snapshot`. That also means the startup process has to keep the snapshots memory for a long time (P.S. currently it still does, but I have a `XXX` message to remind myself to clean them up as soon as snapshot is exported :) ).

> We would need to have another manifest file to store the restore point -> snapshot mapping

I believe this suggestion based on the point that we use the upstream file name format. So let's sort out that point first.

Basically I feel that the upstream format is incompatible with the "persistent" notion, because all the backend ID, counter stuff are not really persistent. So a new format is needed nonetheless, perhaps we can just have `persistent_<snapshot_name>`, where snapshot_name can be restore point name. Thoughts?

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -13713,3 +13769,31 @@ XLogRequestWalReceiverReply(void)
...
+ */
+void
+InitRestorePointInfo(void)
+{
+	HASHCTL		info;
+	long		init_table_size,
+				max_table_size;
+
+	init_table_size = 16;
+	max_table_size = 64;
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/backend/access/transam/xlog.c:13794]: This is somewhat related to the decision to make `gp_hot_standby_snapshot_mode` PGC_POSTMASTER. But I guess we can have a GUC for it regardless. Will do.

**Comment by soumyadeep2007** [comment]
> > We would like to continue to use the same file format and directory location (no prefix, string manipulation etc)
> 
> We cannot use the current format, as (1) there's no valid xid for the startup process who creates the snapshot; (2) the name won't be unique across server restart (as backend ID would repeat and the snapshot counter starts from 1 again). So some kind of special format is needed, but we can discuss more how it should look like.

That is problematic. Aargh, most of my comments were based on the assumption that we can use the upstream format.

File: src/backend/utils/misc/guc_gp.c [hunk 2]
```diff
@@ -5347,6 +5376,38 @@ assign_gp_default_storage_options(const char *newval, void *extra)
...
+		gp_hot_standby_snapshot_mode = HS_SNAPSHOT_UNSYNC;
+	else if (strcmp(newval, HS_SNAPSHOT_RESTOREPOINT_STR) == 0)
+		gp_hot_standby_snapshot_mode = HS_SNAPSHOT_RESTOREPOINT;
+	else
+		/* option not valid, but we don't want panic either */
+		ereport(WARNING,
+			(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid gp_hot_standby_snapshot_mode value"),
+				 errhint("Use one of these: \"unsync\", \"restorepoint\".")));
+}
```
Comment (soumyadeep2007) [inline src/backend/utils/misc/guc_gp.c:5409]: We can avoid all this if we list the GUC under `ConfigureNamesEnum`. See an example: `bytea_output`

File: src/backend/utils/init/postinit.c [hunk 3]
```diff
@@ -838,7 +841,20 @@ InitPostgres(const char *in_dbname, Oid dboid, const char *username,
...
+		/*
+		 * GPDB: on hot standby, we might need to rely on restore points
+		 * to create snapshot (HS_SNAPSHOT_RESTOREPOINT). If no restore 
+		 * point is provided, we don't serve queries but we still want to
+		 * be able to be connected (e.g. so restore point can be provided).
+		 * So always temporarily use unsync snapshot to allow connection.
+		 * We are not using this snapshot anyway, as mentioned by comments above.
+		 */
+		save = gp_hot_standby_snapshot_mode;
+		gp_hot_standby_snapshot_mode = HS_SNAPSHOT_UNSYNC;
```
Comment (soumyadeep2007) [inline src/backend/utils/init/postinit.c:853]: I'm not quite sure I follow the save - call - reset here.

File: src/backend/utils/time/snapmgr.c [hunk 4]
```diff
@@ -299,6 +299,105 @@ SnapMgrInit(void)
...
+	 * Now get the restore point name. 
+	 * For QD, it is either specified by GUC or just using the latest restore point.
+	 * For QE, it is passed via distributed snapshot dispatched from QD.
+	 */
+	memset(rpname, 0, MAXFNAMELEN);
+	if (IS_HOT_STANDBY_QD())
+	{
+		if (!gp_hot_standby_snapshot_restore_point_name)
+		{
+			ereport(DEBUG1,
```
Comment (soumyadeep2007) [inline src/backend/utils/time/snapmgr.c:348]: What is the guarantee that shmLatestRpName is set here?

I don't think we should go into the business of using the latest snapshot if the GUC is not set. We want this to be very explicit.

I think we should throw a FATAL error here if `gp_hot_standby_snapshot_restore_point_name` is not set but `gp_hot_standby_snapshot_mode == HS_SNAPSHOT_RESTOREPOINT`.

This can be similar to:
```
			ereport(FATAL,
					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
					 errmsg("database \"%s\" is not currently accepting connections",
							name)));
```

We should be doing this check much earlier, in `InitPostgres()`, where we should FATAL out. In this function, we should just Assert that the GUC's value is set to something sane.

File: src/backend/utils/misc/guc_gp.c [hunk 5]
```diff
@@ -4774,6 +4781,28 @@ struct config_string ConfigureNamesString_gp[] =
 		NULL, NULL, NULL
 	},
 
+	{
+		{"gp_hot_standby_snapshot_mode", PGC_USERSET, CUSTOM_OPTIONS,
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/utils/misc/guc_gp.c:4785]: I doubt that users would actually want to switch between modes like that on a live cluster. Either restore points are being used or not. We can always enhance it on request.

Restore points can also be used on the full cluster.

I also feel uncomfortable with `gp_hot_standby_snapshot_restore_point_name` being `PGC_USERSET`. Like we can kick off a HS transaction and in the same transaction change it? That is confusing. It feels like more of a PGC_SIGHUP to me.

File: src/backend/utils/time/snapmgr.c [hunk 6]
```diff
@@ -1367,11 +1493,15 @@ ExportSnapshot(Snapshot snapshot)
...
+	 * Generate file path for the snapshot if not given.  We start numbering
+	 * of snapshots inside the transaction from 1.
 	 */
-	snprintf(path, sizeof(path), SNAPSHOT_EXPORT_DIR "/%08X-%08X-%d",
-			 MyProc->backendId, MyProc->lxid, list_length(exportedSnapshots) + 1);
+	if (!snapshot_name)
+		snprintf(path, sizeof(path), SNAPSHOT_EXPORT_DIR "/%08X-%08X-%d",
+				 MyProc->backendId, MyProc->lxid, list_length(exportedSnapshots) + 1);
+	else
+		snprintf(path, sizeof(path), SNAPSHOT_EXPORT_DIR "/%s",
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/backend/utils/time/snapmgr.c:1503]: Still thinking about this. Creating subdirectory during startup would be a net new thing we would be doing, compared to parsing filename which really doesn't add much cost (plus it's just one-time). It might be tricky for the `AllocateDir` and `ReadDirExtended` to work with subdirectory too.

> Makes them easy to avoid too in DeleteAllExportedSnapshotFiles (which we can just modify to exclude that dir) - we don't need another similar function

Technically there's no problem just skip snapshots that have special name in `DeleteAllExportedSnapshotFiles` too. The reason for adding a new function was that I wanted the non-hot-standby nodes to run the old function (i.e. even a standby node w/ `hot_standby=off`, there's no point keeping the RP snapshots). But I think we can just check `ArchiveRecoveryRequested && EnableHotStandby` inside `DeleteAllExportedSnapshotFiles()` to do the same job instead of adding a new function - will do. So the decision about having a dedicated directory is not related.

File: src/backend/utils/time/snapmgr.c (hunk 4 above)
Comment (huansong) [reply src/backend/utils/time/snapmgr.c]: When `shmLatestRpName` is NULL, we would error out later at `cannot find snapshot to use for restore point`.

I think eventually we should provide the option to use the latest RP, either by setting to `NULL` or something more explicit like `latest`. It would be very common for a user in wanting to use whatever the latest they can use. The orchestration between primary and standby would be too annoying. However, I do agree that at this point, allowing latest isn't too helpful, because (1) DR is going to always set it; (2) w/o DR, we don't have a way to make sure every segment has replayed the latest RP, so query can error out easily if one segment lags behind. We should use the *last-universally-replayed* RP instead of the *last-QD-replayed* RP. So I've removed that shmLatestRpName business.

File: src/backend/utils/misc/guc_gp.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/utils/misc/guc_gp.c]: I just don't see the GUC being set by any user running read queries. Its more like an administrative/tool-automated task requiring higher privileges. `PGC_SUSET` seems like the best choice to me.

File: src/backend/cdb/cdbdtxcontextinfo.c [hunk 7]
```diff
@@ -72,13 +72,17 @@ DtxContextInfo_CreateOnCoordinator(DtxContextInfo *dtxContextInfo, bool inCursor
...
 	{
-		DistributedSnapshot_Copy(&dtxContextInfo->distributedSnapshot,
-								 &snapshot->distribSnapshotWithLocalMapping.ds);
-		dtxContextInfo->haveDistributedSnapshot = true;
+		dtxContextInfo->snapshotMode = snapshot->snapshotMode;
+		if (snapshot->snapshotMode == GP_SNAPSHOT_MODE_DISTRIBUTED)
+			DistributedSnapshot_Copy(&dtxContextInfo->distributedSnapshot,
+									 &snapshot->distribSnapshotWithLocalMapping.ds);
+		else if (snapshot->snapshotMode == GP_SNAPSHOT_MODE_RESTOREPOINT)
+			strncpy(&dtxContextInfo->rpname[0], &snapshot->rpname[0], MAXFNAMELEN);
```
Comment (soumyadeep2007) [inline src/backend/cdb/cdbdtxcontextinfo.c:82]: Use `StrNCpy` as it guarantees that the result string is null-terminated.

File: src/backend/storage/ipc/standby.c [hunk 8]
```diff
@@ -316,6 +317,60 @@ ResolveRecoveryConflictWithVirtualXIDs(VirtualTransactionId *waitlist,
...
+	RestorePointInfo 	rp;
+	bool 			found;
+	char 			snapshotname[MAXFNAMELEN + sizeof(RP_SNAPSHOT_PREFIX)];
+	List 			*files = NULL;
+
+	hash_seq_init(&hash_seq, RestorePointHash);
+
+	/*
+	 * Lock to protect the access to RestorePointHash, until we removed the 
+	 * entry *and* the cooresponding snapshot files.
```
Comment (soumyadeep2007) [inline src/backend/storage/ipc/standby.c:337]: s/cooresponding/corresponding

File: src/backend/storage/ipc/standby.c [hunk 9]
```diff
@@ -316,6 +317,60 @@ ResolveRecoveryConflictWithVirtualXIDs(VirtualTransactionId *waitlist,
...
+				hash_seq_term(&hash_seq);
+				ereport(WARNING,
+							errmsg("Cannot remove RP %s. Entry not existed. ", rp->rpname));
+				break;
+			}
+			sprintf(snapshotname, "%s%s",
+						RP_SNAPSHOT_PREFIX,
+						rp->rpname);
+			files = lappend(files, pstrdup(snapshotname));
+			ereport(DEBUG1,
```
Comment (soumyadeep2007) [inline src/backend/storage/ipc/standby.c:364]: Should be under Debug_print_full_dtm, LOG instead.

File: src/backend/storage/ipc/standby.c [hunk 10]
```diff
@@ -316,6 +317,60 @@ ResolveRecoveryConflictWithVirtualXIDs(VirtualTransactionId *waitlist,
...
+			hash_search(RestorePointHash, rp->rpname, HASH_REMOVE, &found);
+			if (!found)
+			{
+				/*
+				 * This shouldn't happen, but we don't want panic in
+				 * startup process, just log and exit gracefuly.
+				 */
+				hash_seq_term(&hash_seq);
+				ereport(WARNING,
+							errmsg("Cannot remove RP %s. Entry not existed. ", rp->rpname));
```
Comment (soumyadeep2007) [inline src/backend/storage/ipc/standby.c:357]: `could not find restore point \"%s\" to invalidate`

File: src/backend/utils/misc/guc_gp.c [hunk 11]
```diff
@@ -4431,6 +4441,16 @@ struct config_int ConfigureNamesInt_gp[] =
 	},
 #endif
 
+	{
+		{"gp_max_restore_point_snapshots", PGC_POSTMASTER, CUSTOM_OPTIONS,
```
Comment (soumyadeep2007) [inline src/backend/utils/misc/guc_gp.c:4445]: The messaging around this needs a bit more thought.

Also, this might be influenced by our decisions on the cleanup mechanism.

File: src/backend/storage/ipc/standby.c [hunk 12]
```diff
@@ -316,6 +317,60 @@ ResolveRecoveryConflictWithVirtualXIDs(VirtualTransactionId *waitlist,
...
+	LWLockAcquire(RestorePointHashLock, LW_EXCLUSIVE);
+
+	while ((rp = hash_seq_search(&hash_seq)) != NULL)
+	{
+		Assert(TransactionIdIsValid(rp->xmin));
+
+		if (!TransactionIdIsValid(latestRemovedXid) ||
+			!TransactionIdFollows(rp->xmin, latestRemovedXid))
+		{
+			hash_search(RestorePointHash, rp->rpname, HASH_REMOVE, &found);
```
Comment (soumyadeep2007) [inline src/backend/storage/ipc/standby.c:348]: This is currently the only way to remove these restore points.

We have to think of a mechanism to automatically remove these restore points at a regular cadence. We don't want the hash table nor the number of files on disk to grow to some high number (imagine automated restore points every few minutes with a high cluster uptime and no conflicts ..  can happen for non-regular vacuum schedules etc). This can cause this loop to take much longer etc.

This also segues in to the question: should we have a hard limit on the number of RP based restore points, i.e. make `gp_max_restore_point_snapshots` a hard limit?)

One of the mechanisms could be to clean "old" restore points at checkpoints. The other can be by making it DBA/automation-tool-driven with the help of an explicit UDF etc.

File: src/backend/utils/time/snapmgr.c [hunk 13]
```diff
@@ -299,6 +299,104 @@ SnapMgrInit(void)
...
+
+	/* Now find the restore point hash entry, and import the snapshot */
+	LWLockAcquire(RestorePointHashLock, LW_SHARED);
+	rp = hash_search(RestorePointHash, rpname, HASH_FIND, &found);
+	if (!found)
+		ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("cannot find snapshot to use for restore point \"%s\"",
+								rpname),
+					errhint("The restore point does not exist or the snapshot has been invalidated.")));
```
Comment (soumyadeep2007) [inline src/backend/utils/time/snapmgr.c:372]: The -> the

**Review by soumyadeep2007 — COMMENTED** [review]
Another pass.

General comment: we should add debug logs under Debug_print_full_dtm, and look for the opportunity to add more such logs, in core routines.
